File: tests/rust/game/panics.rs
//! The panic pins: rust's error-handling story at the FFI boundary.
//! A panic MUST NOT unwind across the C seam (instant UB / abort) —
//! the glue's catch_unwind at every entry point and around every hook
//! is what these scripts exist to trip, tick after tick.

use crate::labelle::{self, Script};

/// `update` panics every tick: logged EVERY time, never evicted (state
/// is intact — the author gets the report until it's fixed), siblings
/// unaffected. The embedded-family analog of a lua error() in update.
pub struct Exploder;

impl Script for Exploder {
    fn update(&mut self, _dt: f32) {
        panic!("boom on tick");
    }
}

/// `init` panics: the script is EVICTED — its update/deinit must never
/// run against half-initialized state. Mirrors the lua suite's
/// "a script whose init() errors is evicted from update and deinit".
pub struct BadInit;

impl Script for BadInit {
    fn init(&mut self) {
        panic!("bad_init boom");
    }

    fn update(&mut self, _dt: f32) {
        labelle::log("bad_init update ran");
    }

    fn deinit(&mut self) {
        labelle::log("bad_init deinit ran");
    }
}

File: tests/go/game/panics.go
// The panic-containment doubles: every hook panic must be recovered at
// the FFI boundary by the shipped glue (a Go panic unwinding out of a
// cgo export would kill the test process — surviving these scenarios
// IS the pin).
package game

import "labelle"

// Exploder panics in every Update — logged every tick, never evicted,
// siblings unaffected.
type Exploder struct{ labelle.NoopScript }

func (Exploder) Update(float32) {
	panic("boom on tick")
}

// BadInit panics in Init — logged and EVICTED: Update/Deinit must
// never run on half-initialized state.
type BadInit struct{ labelle.NoopScript }

func (BadInit) Init() {
	panic("bad_init boom")
}

func (BadInit) Update(float32) {
	labelle.Log("bad_init update ran")
}

func (BadInit) Deinit() {
	labelle.Log("bad_init deinit ran")
}
